examples/object_detection/ios/ObjectDetector/ViewContollers/MediaLibraryViewController.swift

```diff
@@ -0,0 +1,358 @@
+          }
+          return
+        }
+          
+        DispatchQueue.main.async {
+          weakSelf.hideProgressView()
+          weakSelf.draw(
+            detections: objectDetectorResult.detections,
+            originalImageSize: image.size,
+            andOrientation: image.imageOrientation)
+        }
+      }
+    default:
+      break
+    }
+  }
+  
+  func clearAndInitializeObjectDetectorService(runningMode: RunningMode) {
+    objectDetectorService = nil
+    switch runningMode {
+      case .image:
+        objectDetectorService = ObjectDetectorService
+          .stillImageDetectorService(
+            model: DetectorMetadata.sharedInstance.model,
+            maxResults: DetectorMetadata.sharedInstance.maxResults,
+            scoreThreshold: DetectorMetadata.sharedInstance.scoreThreshold)
+      case .video:
+        objectDetectorService = ObjectDetectorService
+          .videoObjectDetectorService(
+            model: DetectorMetadata.sharedInstance.model,
+            maxResults: DetectorMetadata.sharedInstance.maxResults,
+            scoreThreshold: DetectorMetadata.sharedInstance.scoreThreshold,
+            videoDelegate: self)
+      default:
+        break;
+    }
+  }
+  
+  private func playVideo(mediaURL: URL, resultBundle: ResultBundle?) {
+    playVideo(asset: AVAsset(url: mediaURL))
+    playerTimeObserverToken = playerViewController?.player?.addPeriodicTimeObserver(
+      forInterval: CMTime(value: Constants.inferenceTimeIntervalMs,
+                          timescale: Int32(Constants.kMilliSeconds)),
+      queue: DispatchQueue(label: "timeObserverQueue", qos: .userInteractive),
+      using: { [weak self] (time: CMTime) in
+        DispatchQueue.main.async {
+          let index =
+            Int(CMTimeGetSeconds(time) * Double(Constants.kMilliSeconds) / Double(Constants.inferenceTimeIntervalMs))
+          guard
+                let weakSelf = self,
+                let resultBundle = resultBundle,
+                index < resultBundle.objectDetectorResults.count,
+                let objectDetectorResult = resultBundle.objectDetectorResults[index] else {
+            return
+          }
+          weakSelf.draw(
+            detections: objectDetectorResult.detections,
+            originalImageSize: resultBundle.size,
+            andOrientation: .up)
+        }
+    })
+  }
+  
+  private func playVideo(asset: AVAsset) {
+    if playerViewController == nil {
+      let playerViewController = AVPlayerViewController()
+      self.playerViewController = playerViewController
+    }
+    
+    let playerItem = AVPlayerItem(asset: asset)
+    if let player = playerViewController?.player {
+      player.replaceCurrentItem(with: playerItem)
+    }
+    else {
+      playerViewController?.player = AVPlayer(playerItem: playerItem)
+    }
+    
+    playerViewController?.showsPlaybackControls = false
+    addPlayerViewControllerAsChild()
+    
+    NotificationCenter.default
```

Comment:
KVO for DetectorMetadata is fine because multiple classes needs to listen to it.

For this class, what about using a standard Timer to display the bounding boxes for simplicity? IIRC the original sample used this approach. We don't allow pause/resume the video player so the logic doesn't need to be tightly coupled.
- The timer ticks periodically right after the video starts
- If the timer's time exceed the video type, then clear all object bounding boxes. The lag time between when the video ends to clearing the bounding boxes will be 300ms at max, which I think is acceptable for the sake of code simplicity.

Reply:
I tried it with a timer. There are other notifications we will end up adding to handle display when app enters background and foreground in this case.  When the app goes to the background, the playback is automatically paused. iOS does not resume this when the app comes back to the foreground ( can't change this behaviour).  If we use the `addPeriodicTimeObserver` on avplayer for drawing, this callback also gets paused automatically when player is paused by the OS. But if we use our own timer, we will have to register for the `app did become inactive notification `to invalidate the timer. Otherwise when you bring it back to foreground, you will see a paused player display on which boxes are being drawn periodically because the NSTimer is still active.

We can void the avplayeritem did play to end notification by using the calculation you outlined even if we retain the player's periodic time observer.

Bottom line is that either we need the player's time observer or the notification when the app enters background. 

FYI, whether we use, the player or the periodic time observer if the video playback should be resumed when the app enters foreground, we will have to register for the app did become active notification. But since this is an example app, we can choose to not handle this.

Reply:
Thanks for the thoughtful consideration. I looked at the code one more time and I think it's concise enough as it. Let's keep your current implementation.

nit: I'd suggest adding some comments to playVideo method to explain what you do there, so the developers who read the code can quickly get an idea of what you're doing in the method.